From aa5276c00c625cb7daabe2d9e6d4a28bb4041a5f Mon Sep 17 00:00:00 2001
From: thampan <joby.thampan@intel.com>
Date: Thu, 14 Nov 2019 15:39:31 +0800
Subject: [PATCH] DRV_DPM_SW-86: DPM PMAC Egress Config setting corrected

    - Bug Fix for DPM not setting correct information for Mixed Mode, flag mode and others
    - GSWIP Pmac Global Get bug fix
---
 drivers/net/datapath/dpm/gswip31/datapath_gswip.c  | 61 ++++++++++++++++------
 .../net/ethernet/lantiq/switch-api/gsw_flow_core.c | 25 +++++++--
 2 files changed, 67 insertions(+), 19 deletions(-)

diff --git a/drivers/net/datapath/dpm/gswip31/datapath_gswip.c b/drivers/net/datapath/dpm/gswip31/datapath_gswip.c
index 6f5927e0eaac..d6a0df23d704 100644
--- a/drivers/net/datapath/dpm/gswip31/datapath_gswip.c
+++ b/drivers/net/datapath/dpm/gswip31/datapath_gswip.c
@@ -86,12 +86,13 @@ int dp_pmac_set_31(int inst, u32 port, dp_pmac_cfg_t *pmac_cfg)
 {
 	GSW_PMAC_Eg_Cfg_t egcfg;
 	GSW_PMAC_Ig_Cfg_t igcfg;
+	GSW_PMAC_Glbl_Cfg_t glblCfg = {0};
 	int i, j, k;
 	u32 flag = 0;
 	cbm_dq_port_res_t dqport;
 	s32 ret;
 	u32 pmacid;
-	struct core_ops *gswr_r;
+	struct core_ops *gsw_ops;
 
 	if (!pmac_cfg || !port) {
 		pr_err("dp_pmac_set:wrong parameter(pmac_cfg/port NULL)\n");
@@ -107,7 +108,7 @@ int dp_pmac_set_31(int inst, u32 port, dp_pmac_cfg_t *pmac_cfg)
 	memset(&dqport, 0, sizeof(cbm_dq_port_res_t));
 
 	/* Get GSWIP device handler */
-	gswr_r = dp_port_prop[inst].ops[0];
+	gsw_ops = dp_port_prop[inst].ops[0];
 
 	/*set ingress port via DMA tx channel */
 	if (pmac_cfg->ig_pmac_flags) {
@@ -124,11 +125,11 @@ int dp_pmac_set_31(int inst, u32 port, dp_pmac_cfg_t *pmac_cfg)
 		for (i = 0; i < dqport.num_deq_ports; i++) {
 			igcfg.nPmacId = pmacid;
 			igcfg.nTxDmaChanId = dqport.deq_info[i].dma_tx_chan;
-			if (igcfg.nTxDmaChanId == (u8)-1) {
+			if (igcfg.nTxDmaChanId == (u8) - 1) {
 				igcfg.nTxDmaChanId =
 					pmac_cfg->ig_pmac.tx_dma_chan;
 			}
-			gswr_r->gsw_pmac_ops.Pmac_Ig_CfgGet(gswr_r, &igcfg);
+			gsw_ops->gsw_pmac_ops.Pmac_Ig_CfgGet(gsw_ops, &igcfg);
 
 			/*update igcfg and write back to gsw */
 			if (pmac_cfg->ig_pmac_flags & IG_PMAC_F_ERR_DISC)
@@ -220,37 +221,67 @@ int dp_pmac_set_31(int inst, u32 port, dp_pmac_cfg_t *pmac_cfg)
 			DP_DEBUG(DP_DBG_FLAG_DBG, "igcfg.defPmacHdr=");
 
 			for (k = 0;
-					k <
-					sizeof(igcfg.defPmacHdr) /
-					sizeof(igcfg.defPmacHdr[0]); k++)
+			     k <
+			     sizeof(igcfg.defPmacHdr) /
+			     sizeof(igcfg.defPmacHdr[0]); k++)
 				DP_DEBUG(DP_DBG_FLAG_DBG, "0x%x ",
 					 igcfg.defPmacHdr[k]);
 
 			DP_DEBUG(DP_DBG_FLAG_DBG, "\n");
 
-			gswr_r->gsw_pmac_ops.Pmac_Ig_CfgSet(gswr_r, &igcfg);
+			gsw_ops->gsw_pmac_ops.Pmac_Ig_CfgSet(gsw_ops, &igcfg);
 		}
 
-			kfree(dqport.deq_info);
+		kfree(dqport.deq_info);
 	}
 
 	/*set egress port via pmac port id */
 	if (!pmac_cfg->eg_pmac_flags)
 		return 0;
 
+	ret = gsw_ops->gsw_pmac_ops.Pmac_Gbl_CfgGet(gsw_ops, &glblCfg);
+	if (ret) {
+		pr_err("Pmac_Gbl_CfgGet failed\n");
+		return -1;
+	}
+
 	for (i = 0; i <= 15; i++) {	/*traffic class */
 		for (j = 0; j <= 3; j++) {	/* flow */
 			/*read back egcfg first from gsw */
 			memset(&egcfg, 0, sizeof(GSW_PMAC_Eg_Cfg_t));
 			egcfg.nPmacId = pmacid;
 			egcfg.nDestPortId = port;
-			egcfg.nTrafficClass = i;
 			egcfg.nFlowIDMsb = j;
 			egcfg.nBslTrafficClass = i;
-			egcfg.bProcFlagsSelect = 1;
+
+			switch (glblCfg.eProcFlagsEgCfg) {
+			/* Use reduced traffic class (saturated to 3)
+			 * and flags (MPE1, MPE2)
+			 */
+			case GSW_PMAC_PROC_FLAGS_MIX:
+				egcfg.bProcFlagsSelect = 1;
+				egcfg.nTrafficClass = (i & 0xC) >> 2;
+				egcfg.bMpe2Flag = (i & 0x2) >> 1;
+				egcfg.bMpe1Flag = (i & 0x1);
+				break;
+			/* Use flags (MPE1, MPE2, DEC, ENC) */
+			case GSW_PMAC_PROC_FLAGS_FLAG:
+				egcfg.bProcFlagsSelect = 1;
+				egcfg.bDecFlag = (i & 0x8) >> 3;
+				egcfg.bEncFlag = (i & 0x4) >> 2;
+				egcfg.bMpe2Flag = (i & 0x2) >> 1;
+				egcfg.bMpe1Flag = (i & 0x1);
+				break;
+			/* Use traffic class as full 4 bits */
+			default:
+				egcfg.bProcFlagsSelect = 0;
+				egcfg.nTrafficClass = i;
+				break;
+			}
+
 			DP_DEBUG(DP_DBG_FLAG_DBG, "bProcFlagsSelect=%u\n",
 				 egcfg.bProcFlagsSelect);
-			gswr_r->gsw_pmac_ops.Pmac_Eg_CfgGet(gswr_r, &egcfg);
+			gsw_ops->gsw_pmac_ops.Pmac_Eg_CfgGet(gsw_ops, &egcfg);
 
 			/*update egcfg and write back to gsw */
 			if (pmac_cfg->eg_pmac_flags & EG_PMAC_F_FCS)
@@ -259,7 +290,7 @@ int dp_pmac_set_31(int inst, u32 port, dp_pmac_cfg_t *pmac_cfg)
 			if (pmac_cfg->eg_pmac_flags & EG_PMAC_F_L2HDR_RM) {
 				egcfg.bRemL2Hdr = pmac_cfg->eg_pmac.rm_l2hdr;
 				egcfg.numBytesRem =
-				    pmac_cfg->eg_pmac.num_l2hdr_bytes_rm;
+					pmac_cfg->eg_pmac.num_l2hdr_bytes_rm;
 			}
 
 			if (pmac_cfg->eg_pmac_flags & EG_PMAC_F_PMAC)
@@ -267,7 +298,7 @@ int dp_pmac_set_31(int inst, u32 port, dp_pmac_cfg_t *pmac_cfg)
 
 			if (pmac_cfg->eg_pmac_flags & EG_PMAC_F_RXID)
 				egcfg.nRxDmaChanId =
-				    pmac_cfg->eg_pmac.rx_dma_chan;
+					pmac_cfg->eg_pmac.rx_dma_chan;
 
 			if (pmac_cfg->eg_pmac_flags & EG_PMAC_F_REDIREN)
 				egcfg.bRedirEnable = pmac_cfg->eg_pmac.redir;
@@ -380,7 +411,7 @@ int dp_pmac_set_31(int inst, u32 port, dp_pmac_cfg_t *pmac_cfg)
 					 egcfg.bMpe2Flag);
 			}
 #endif
-			gswr_r->gsw_pmac_ops.Pmac_Eg_CfgSet(gswr_r, &egcfg);
+			gsw_ops->gsw_pmac_ops.Pmac_Eg_CfgSet(gsw_ops, &egcfg);
 		}
 	}
 
diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
index 1a782e2490c0..2b4007babda4 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
@@ -16299,6 +16299,7 @@ GSW_return_t GSW_PMAC_GLBL_CfgSet(void *cdev, GSW_PMAC_Glbl_Cfg_t *parm)
 			case GSW_PMAC_PROC_FLAGS_FLAG:
 				pmac_ctrl4 = gsw_field_w32(pmac_ctrl4, PMAC_CTRL_4_FLAGEN_SHIFT,
 							   PMAC_CTRL_4_GSWIP3_1_FLAGEN_SIZE, 1);
+
 				break;
 
 			case GSW_PMAC_PROC_FLAGS_MIX:
@@ -16359,7 +16360,8 @@ GSW_return_t GSW_PMAC_GLBL_CfgGet(void *cdev, GSW_PMAC_Glbl_Cfg_t *parm)
 	    pmac_ctrl4_off,
 	    pmac_bsl1_off,
 	    pmac_bsl2_off,
-	    pmac_bsl3_off;
+	    pmac_bsl3_off,
+	    val;
 	u32 ret;
 
 	if (gswdev == NULL) {
@@ -16435,10 +16437,25 @@ GSW_return_t GSW_PMAC_GLBL_CfgGet(void *cdev, GSW_PMAC_Glbl_Cfg_t *parm)
 			gsw_r32_raw(cdev, pmac_ctrl1_off, &regval);
 
 			parm->nMaxJumboLen = regval;
+			val = gsw_field_r32(pmac_ctrl4, PMAC_CTRL_4_FLAGEN_SHIFT,
+					    PMAC_CTRL_4_GSWIP3_1_FLAGEN_SIZE);
+
+			if (val)
+				parm->bProcFlagsEgCfgEna = 1;
 
-			parm->eProcFlagsEgCfg = gsw_field_r32(pmac_ctrl4,
-							      PMAC_CTRL_4_FLAGEN_SHIFT,
-							      PMAC_CTRL_4_GSWIP3_1_FLAGEN_SIZE);
+			switch (val) {
+			case 0:
+				parm->eProcFlagsEgCfg = GSW_PMAC_PROC_FLAGS_TC;
+				break;
+			case 1:
+				parm->eProcFlagsEgCfg = GSW_PMAC_PROC_FLAGS_FLAG;
+				break;
+			case 2:
+				parm->eProcFlagsEgCfg = GSW_PMAC_PROC_FLAGS_MIX;
+				break;
+			default:
+				break;
+			}
 
 			gsw_r32(cdev,
 				pmac_bsl1_off,
